server/src-lib/Hasura/GraphQL/Schema.hs

```diff
@@ -1,960 +1,598 @@
+
+  functionConnectionSelectParsers <-
+    for allFunctions $ \function -> runMaybeT do
+      let returnTable = fiReturnType function
+          functionName = fiName function
+      pkeyColumns <- MaybeT $ (^? tiCoreInfo.tciPrimaryKey._Just.pkColumns)
+                     <$> askTableInfo returnTable
+      selectPerms <- MaybeT $ tableSelectPermissions returnTable
+      displayName <- qualifiedObjectToName functionName
+      let fieldName = displayName <> $$(G.litName "_connection")
+          fieldDesc = Just $ G.Description $ "execute function " <> functionName
+                      <<> " which returns " <>> returnTable
+      lift $ selectFunctionConnection function fieldName fieldDesc pkeyColumns selectPerms
+
+  pure $ map ((RFDB . QDBConnection) <$>) $ catMaybes $
+         tableConnectionSelectParsers <> functionConnectionSelectParsers
+
+-- | Parse query-type GraphQL requests without introspection
+query
+  :: forall m n r
+   . (MonadSchema n m, MonadTableInfo r m, MonadRole r m, Has QueryContext r)
+  => G.Name
+  -> HashSet QualifiedTable
   -> [FunctionInfo]
-  -> Maybe ViewInfo
-  -> TableConfig
-  -> RoleName
-  -> RolePermInfo
-  -> m (TyAgg, RootFields, InsCtxMap)
-mkGCtxRole tableCache tn descM fields primaryKey constraints funcs viM tabConfigM role permInfo = do
-  selPermM <- mapM (getSelPerm tableCache fields role) $ _permSel permInfo
-  tabInsInfoM <- forM (_permIns permInfo) $ \ipi -> do
-    ctx <- mkInsCtx role tableCache fields ipi $ _permUpd permInfo
-    let permCols = flip getColInfos allCols $ Set.toList $ ipiCols ipi
-    return (ctx, (permCols, icRelations ctx))
-  let insPermM = snd <$> tabInsInfoM
-      insCtxM = fst <$> tabInsInfoM
-      updColsM = filterColumnFields . upiCols <$> _permUpd permInfo
-      tyAgg = mkTyAggRole tn descM insPermM selPermM updColsM
-              (void $ _permDel permInfo) primaryKey constraints viM funcs
-      rootFlds = getRootFldsRole tn primaryKey constraints fields funcs
-                 viM permInfo tabConfigM
-      insCtxMap = maybe Map.empty (Map.singleton tn) insCtxM
-  return (tyAgg, rootFlds, insCtxMap)
-  where
-    allCols = getCols fields
-    cols = getValidCols fields
-    filterColumnFields allowedSet =
-      filter ((`Set.member` allowedSet) . pgiColumn) cols
-
-getRootFldsRole
-  :: QualifiedTable
-  -> Maybe (PrimaryKey PGColumnInfo)
-  -> [ConstraintName]
-  -> FieldInfoMap FieldInfo
+  -> [P.FieldParser n (RemoteSchemaInfo, G.Field G.NoFragments P.Variable)]
+  -> [ActionInfo]
+  -> NonObjectTypeMap
+  -> m (Parser 'Output n (OMap.InsOrdHashMap G.Name (QueryRootField UnpreparedValue)))
+query name allTables allFunctions allRemotes allActions nonObjectCustomTypes = do
+  queryFieldsParser <- query' allTables allFunctions allRemotes allActions nonObjectCustomTypes
+  pure $ P.selectionSet name Nothing queryFieldsParser
+    <&> fmap (P.handleTypename (RFRaw . J.String . G.unName))
```

Comment:
`<&> fmap (P.handleTypename (RFRaw . J.String . G.unName))` appears dozen of times; we could probably make it a function / give it a name.